kernel/src/hil/block_storage.rs

```diff
@@ -0,0 +1,268 @@
+            }
+        })
+    }
+
+    pub fn get_length_bytes(&self) -> u32 {
+        self.length_blocks * S
+    }
+}
+
+/// Specifies a storage area with byte granularity.
+pub struct AddressRange {
+    /// Offset from the beginning of the storage device.
+    pub start_address: u64,
+    /// Length of the range.
+    pub length_bytes: u32,
+}
+
+impl AddressRange {
+    pub fn get_end_address(&self) -> u64 {
+        self.start_address + self.length_bytes as u64
+    }
+}
+
+impl<const S: u32> From<Region<S>> for AddressRange {
+    fn from(region: Region<S>) -> Self {
+        let length_bytes = region.get_length_bytes();
+        Self {
+            start_address: region.index.into(),
+            length_bytes,
+        }
+    }
+}
+
+/// Writeable persistent block storage device.
+///
+/// The device is formed from equally-sized storage blocks,
+/// which are arranged one after another, without gaps or overlaps,
+/// to form a linear storage of bytes.
+///
+/// There device is split into blocks in two ways, into:
+/// - erase blocks, which are the smallest unit of space that can be erased
+/// - write blocks, which are the smallest unit of space that can be written
+///
+/// Every byte on the device belongs to exactly one erase block,
+/// and to exactly one write block at the same time.
+///
+/// `W`: The size of a write block in bytes.
+/// `E`: The size of an erase block in bytes.
+pub trait BlockStorage<const W: u32, const E: u32> {
+    /// Read data from flash into a buffer.
+    ///
+    /// This function will read data stored in flash at `range` into `buf`.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `range` exceeds the end of the device, or
+    /// - `buf` is shorter than `range`.
+    ///
+    /// On success, triggers `Client::read_complete` once.
+    /// On failure returns a `ErrorCode` and the buffer passed in.
+    /// If `ErrorCode::NOSUPPORT` is returned then `read_block`
+    /// should be used instead.
+    fn read_range(
+        &self,
+        range: &AddressRange,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Read data from blocks starting at `block`, and into the buffer.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `range` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
+    ///
+    /// On success, triggers `Client::read_complete` once.
+    fn read(
+        &self,
+        region: &Region<W>,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Write data from a buffer to flash.
+    ///
+    /// This function writes the contents of `buf` to memory,
+    /// starting at the chosen `start_block`.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `range` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
+    ///
+    /// This function SHALL NOT prepare the block for writing first.
```

Comment:
```suggestion
    /// This function SHALL NOT prepare the region for writing first.
```
? Or "blocks in region" maybe.

Reply:
I'm wary of this SHALL NOT, because it's an implementation decision. I think you want to say that this method does not any claims of preparation, and that a client that needs a clean block MUST call `prepare_write` first. That leaves open the possibility of multiple writes, while also allowing implementations that defer preparation to this actual method call (e.g., by storing a "prepare first" bit).